src/themes/bootstrap4.js

```diff
@@ -3,114 +3,372 @@ import { AbstractTheme } from '../theme'
 export var bootstrap4Theme = AbstractTheme.extend({
   /* Theme config options that allows changing various aspects of the output */
   options: {
-    'disable_theme_rules': false
+    disable_theme_rules: false,
+    input_size: 'normal', // Size of input and select elements. "small", "normal", "large"
+    custom_forms: false // use twbs custom form stylings
   },
   /* Custom stylesheet rules. format: "selector" : "CSS rules" */
   rules: {
-    'div[data-schemaid="root"]:after': 'position:relative;color:red;margin:10px 0;font-weight:600;display:block;width:100%;text-align:center;content:"This is an old JSON-Editor 1.x Theme and might not display elements correctly when used with the 2.x version"'
+    '.jsoneditor-twbs4-text-button': 'background: none;padding: 0;border: 0',
+    'td>.form-group': 'margin-bottom: 0',
+    '.json-editor-btn-upload': 'margin-top: 1rem'
   },
+
   getSelectInput: function (options, multiple) {
     var el = this._super(options)
     el.classList.add('form-control')
-    // el.style.width = 'auto';
+
+    if (this.options.custom_forms === false) {
+      if (this.options.input_size === 'small') el.classList.add('form-control-sm')
+      if (this.options.input_size === 'large') el.classList.add('form-control-lg')
+    } else {
+      el.classList.remove('form-control')
+      el.classList.add('custom-select')
+      if (this.options.input_size === 'small') el.classList.add('custom-select-sm')
+      if (this.options.input_size === 'large') el.classList.add('custom-select-lg')
+    }
+
     return el
   },
+
   setGridColumnSize: function (el, size, offset) {
     el.classList.add('col-md-' + size)
+
     if (offset) {
       el.classList.add('offset-md-' + offset)
     }
   },
+
   afterInputReady: function (input) {
     if (input.controlgroup) return
-    input.controlgroup = this.closest(input, '.form-group')
-    if (this.closest(input, '.compact')) {
-      input.controlgroup.style.marginBottom = 0
+
+    // set id/for
+    // is not working for: [type=file], [type=checkbox]
```

Comment:
This makes sense, as the checkbox and file editors have not been refactored yet. So they might not be theme friendly like the other editors/types.